### DISCUSSION TITLE
emmet and pydal

### DISCUSSION THREAD

**Discussion by sugizo** [emmett-framework/emmett#304] [Q&A]
```
$ python3 server.py 
Traceback (most recent call last):
  File "server.py", line 1, in <module>
    from music import *
  File "/Users/sugizo/Downloads/emmet/music/__init__.py", line 1, in <module>
    from .db_schema.db_schema_1_instrument import db
  File "/Users/sugizo/Downloads/emmet/music/db_schema/db_schema_1_instrument.py", line 1, in <module>
    from ..models import *
  File "/Users/sugizo/Downloads/emmet/music/models.py", line 2, in <module>
    from pydal.validators import *
ModuleNotFoundError: No module named 'pydal.validators'

$ pip3 install -U pydal
Requirement already satisfied: pydal in /Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages (17.3)
Collecting pydal
  Using cached pydal-20200910.1-py2.py3-none-any.whl
Installing collected packages: pydal
  Attempting uninstall: pydal
    Found existing installation: pyDAL 17.3
    Uninstalling pyDAL-17.3:
      Successfully uninstalled pyDAL-17.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
emmett 2.1.4 requires pyDAL==17.3, but you have pydal 20200910.1 which is incompatible.
Successfully installed pydal-20200910.1

```

**question**
is emmet will support pydal ?

**n.b.**
current installed pydal work fine with another framework

thanks

**Comment by gi0baro** [comment]
Emmett currently *relies* on pyDAL, but only on the version specified into requirements (17.3 as you noticed from the pip warning).

The thing is pyDAL consist into the lower layer used by Emmett's ORM (https://emmett.sh/docs/2.1.x/orm) which implements its own validation system (https://emmett.sh/docs/2.1.x/validations) then probably using the tools provided by Emmett itself will give you a smooth experience rather than packing up other tools together into your project.

AFAIK, you should be able to use even modern versions of pyDAL (ignoring the dependency error you get from pip), but then you probably loose all the integrations provided by Emmett, like the connection pipes, ORM models and relations, validations, extensions, etc.

Considering all the above, at the moment there are no planned actions to update pyDAL or support newer versions of pyDAL into Emmett, as the ORM part of Emmett evolved into a specific direction and the pyDAL part is used just as a *hidden* layer; in contrast there are plans to actually drop pyDAL dependency in 3.x series, but there's actually no ETA for that.
